### PR TITLE
Evoting improvements

### PR DESCRIPTION
**Pull request by jeffallen**
Protect from panic due to shuffle of nil points.

Added a test mode which makes it easier to do full tests including
finalization.

Added date check on vote cast.

### CODE REVIEW

File: evoting/service/service_test.go [hunk 1]
```diff
@@ -304,6 +307,149 @@ func TestService(t *testing.T) {
...
+			"en": "name in english",
+			"fr": "name in french",
+		},
+		Subtitle: map[string]string{
+			"en": "name in english",
+			"fr": "name in french",
+		},
+		MoreInfoLang: map[string]string{
+			"en": "https://epfl.ch/elections",
+			"fr": "httsp://epfl.ch/votations",
```
Comment (nkcr) [inline evoting/service/service_test.go:417]:
```suggestion
			"fr": "https://epfl.ch/votations",
```

**Comment by ineiti** [comment]
I suppose `TestAfterEnd` should pass, correct?

**Comment by jeffallen** [comment]
This PR is ready to merge, Github and Travis do not agree on the fact that it passed (with one restart of Java tests on one of the 2 builds).

File: evoting/lib/transaction.go [hunk 2]
```diff
@@ -187,9 +187,12 @@ func (t *Transaction) Verify(genesis skipchain.SkipBlockID, s *skipchain.Service
 		}
 		return nil
 	} else if t.Ballot != nil {
-		election, err := GetElection(s, genesis, false, t.User)
-		if err != nil {
-			return err
+		zero := cothority.Suite.Point()
```
Comment (ineiti) [inline evoting/lib/transaction.go:190]: According to this: https://github.com/dedis/kyber/blob/master/group.go#L145, you should compare to `cothority.Suite.Point().Null()`

File: evoting/lib/transaction.go [hunk 3]
```diff
@@ -198,6 +201,21 @@ func (t *Transaction) Verify(genesis skipchain.SkipBlockID, s *skipchain.Service
 			return errors.New("ballot user-id differs from transaction user-id")
 		}
 
+		election, err := GetElection(s, genesis, false, t.User)
+		if err != nil {
+			return err
```
Comment (ineiti) [inline evoting/lib/transaction.go:206]: Please encapsulate errors with `fmt.Errorf` and give some more information what went wrong.

File: evoting/protocol/shuffle.go [hunk 4]
```diff
@@ -130,6 +130,17 @@ func (s *Shuffle) HandlePrompt(prompt MessagePrompt) error {
 		}
 
 		a, b := lib.Split(ballots)
+		// Protect from missing input.
+		for i := range a {
+			if a[i] == nil {
+				a[i] = cothority.Suite.Point()
```
Comment (ineiti) [inline evoting/protocol/shuffle.go:136]: Here, and `b[i]`, is it OK to have a somewhat undefined point? As per the comment above, it's not guaranteed that it's `Null()`.

File: evoting/service/service_test.go [hunk 5]
```diff
@@ -304,6 +307,276 @@ func TestService(t *testing.T) {
...
```
Comment (ineiti) [inline evoting/service/service_test.go:471]: Comment from @Gilthoniel : 
```suggestion
			"fr": "https://epfl.ch/votations",
```

File: evoting/service/service_test.go (hunk 1 above)
Comment (ineiti) [reply evoting/service/service_test.go:544]: above, too, in fact ;)

File: skipchain/skipchain.go [hunk 6]
```diff
@@ -1255,7 +1255,7 @@ func (s *Service) bftForwardLinkLevel0(msg, data []byte) bool {
 			g := func(to []byte, newest *SkipBlock) (out bool) {
 				defer func() {
 					if re := recover(); re != nil {
-						log.Error("Verification function panic: " + re.(string))
+						log.Error(fmt.Sprintf("Verification function panic: %v", re))
```
Comment (ineiti) [inline skipchain/skipchain.go:1258]:
```suggestion
						log.Errorf("Verification function panic: %v", re)
```

**Pull request merged**